perceptia-one/perceptia/src/components/QuestionView.js

```diff
@@ -112,38 +128,46 @@ class QuestionView extends React.Component {
             questionId: this.state.questions[counter].id,
             question_type: +this.state.questions[counter].q_type,
             paragraph_id: this.state.questions[counter].p_id,
-        })
+            busy: false,
+        });
 
     }
 
     handleAnswerSelected(answer) {
 
-        console.log(answer);//, event.currentTarget.value);
-        //this.setUserAnswer(event.currentTarget.value);
-        if (this.state.counter < this.state.questions.length - 1) {
-            var body = JSON.stringify({
-                "questionID":`${this.state.questionId}`,
-                "selectedAnswer": `${answer}`});
-
-            fetch(`${constants.api.url}/api/v1/anyquiz/questions/grade`, {
-                method: 'post',
-                body: body,
-            })
-            .then( response => response.json())
-            .then(response =>{
+        console.log(answer);
+        
+        if(answer==null){
+            this.setState({grades: this.state.grades.concat([false])});
+            this.setNextQuestion();
+            return;
+        }
 
-                this.setState({grades: this.state.grades.concat([response["result"]])})
+        var body = JSON.stringify({
+            "questionID":`${this.state.questionId}`,
+            "selectedAnswer": `${answer}`});
 
-            }).then(() => this.setNextQuestion())
-            //setTimeout(, 300);
-        } else {
+        fetch(`${constants.api.url}/api/v1/anyquiz/questions/grade`, {
+            method: 'post',
+            body: body,
+        })
+        .then( response => response.json())
+        .then(response =>{
+
+            this.setState({grades: this.state.grades.concat([response["result"]])});
+
+        }).then(() => {
+            if (this.state.counter < this.state.questions.length - 1) {
+                this.setNextQuestion();
+            } else {
+                this.setState({
```

Comment:
1. This particular instance you're commenting on is just me re-organizing Nuo's code to allow the questions to be graded before going to the next/results page. 

2. I had not been made aware of this particular caveat in my previous forays into React. If you can provide a reference that helps introduce and explain it, I'd like to read up on it. My first intuition is that it *seems* like that what I think you're suggesting would cause big challenges for the application programmer (as this seems to infer a RMW issue), and should be handled by the middleware (i.e. react and/or nodejs via it's single-threaded js event loop, https://nodejs.org/de/docs/guides/blocking-vs-non-blocking/, disregarding lower level thread pool, https://nodejs.org/en/docs/guides/dont-block-the-event-loop/, or other explicit methods e.g. Worker Threads, https://nodejs.org/api/worker_threads.html, that react could be making use of unbeknownst to me).

Reply:
Here is it mentioned in the React docs: https://reactjs.org/docs/react-component.html#setstate, see the part about "updater function". I implemented it here if you want an example: https://github.com/uw-thalesians/perceptia-clients/blob/6c8b327e5728989b1218f433d30819dc426ebb67/perceptia-one/perceptia/src/components/QuestionView.js#L99.

Here is an article that talks about it: https://medium.com/@wisecobbler/using-a-function-in-setstate-instead-of-an-object-1f5cfd6e55d1

Reply:
We'll look at moving towards this, but from what I've read in the links provided, I don't want to make these changes last minute before Capstone. There are probably issues with the way we have things right now, but making a change to another method that may also not work the way we all know/expect won't necessarily be any better.